global/scripts/mixtureModel.sh

```diff
@@ -0,0 +1,139 @@
+
+mixtureModel() {
+  local inFile="$1"
+  local outFile="$2"
+  local wbcmd="${3:-wb_command}"
+  local melocmd="${4:-melodic}"
+  local wbshrtctscmd="${5:-wb_shortcuts}"
+
+  # handle inputs
+  if [[ -z "$inFile" || -z "$outFile" ]]; then
+    echo "inFile and outFile arguments are required!" >&2
+    return 1
+  fi
+
+  # check dependencies
+  command -v "$wbcmd" >/dev/null 2>&1 || { echo "workbench_command binary $wbcmd not on path" >&2; return 1; }
+  command -v "$wbshrtctscmd" >/dev/null 2>&1 || { echo "workbench_command binary $wbshrtctscmd not on path" >&2; return 1; }
+  command -v "$melocmd" >/dev/null 2>&1 || { echo "melodic command binary $melocmd not on path" >&2; return 1; }
+
+  local inFile0="$inFile"
+  local outFile0="$outFile"
+  local tDir
+  tDir=$(mktemp -d)
+  local FSLOUTPUTTYPE0="$FSLOUTPUTTYPE"
+  local stat=0
+  local ext
+
+  # convert input if needed
+  if [[ "$inFile0" == *dscalar.nii ]]; then
+    # convert input cifti to nifti
+    inFile="$(mktemp --suffix=.nii)"
+    $wbcmd -cifti-convert -to-nifti "$inFile0" "$inFile" -smaller-dims
+    # local dims
+    dims=$($wbcmd -file-information "$inFile" | grep Dimensions | awk -F': ' '{print $2}')
+    # shellcheck disable=SC2206
+    dims_arr=($dims)
+    for i in "${!dims_arr[@]}"; do
+      if [[ "${dims_arr[$i]}" == "1" && $i -lt 3 ]]; then
+        echo "warning: singleton dimension in converted nifti, melodic may not interpret correctly!" >&2
+        break
+      fi
+    done
+    inFile="${inFile%.nii}"
+  elif [[ "$inFile0" == *.nii ]]; then
+    inFile="${inFile0%.nii}"
+  elif [[ "$inFile0" == *.nii.gz ]]; then
+    inFile="${inFile0%.nii.gz}"
+  else
+    echo "inFile is not a nifti or dscalar cifti?" >&2
+    return 1
+  fi
+
+  # handle output extension
+  if [[ "$outFile0" == *dscalar.nii ]]; then
+    if [[ "$inFile0" != *dscalar.nii ]]; then
+      echo "cifti output only supported for cifti input." >&2
+      return 1
+    fi
+    outFile="${outFile0%.dscalar.nii}"
+    FSLOUTPUTTYPE="NIFTI_GZ"
+    ext="nii.gz"
+  elif [[ "$outFile0" == *.nii ]]; then
+    outFile="${outFile0%.nii}"
+    FSLOUTPUTTYPE="NIFTI"
+    ext=".nii"
+  elif [[ "$outFile0" == *.nii.gz ]]; then
+    FSLOUTPUTTYPE="NIFTI_GZ"
+    ext="nii.gz"
+    outFile="${outFile0%.nii.gz}"
+  else
+    echo "outFile is not a nifti or dscalar cifti?" >&2
+    return 1
+  fi
+
+  # run gaussian mixture modeling with melodic
+  mkdir -p "$tDir"
+  echo "1" > "$tDir/grot"
+  $melocmd -i "$inFile" --ICs="$inFile" --mix="$tDir/grot" -o "$tDir" --Oall --report -v --mmthresh=0
+
+  # check for multiple z-score maps
+  mapCounts=$(find "$tDir/stats/" -name 'thresh_zstat*' -exec bash -c 'fslval "$1" dim4' _ {} \;)
+  maxCount=0
+  for count in $mapCounts; do
+    if (( count > maxCount )); then maxCount=$count; fi
+  done
+  if (( maxCount > 1 )); then
+    echo "warning: At least one component returned two z-score maps (alpha = 0.05 and 0.01)! Using first map only." >&2
+  fi
```

Comment:
You don't need find to do globbing in bash (they aren't in an unpredictable folder name, right?), and you can stop early when you find one.  Can they ever be mismatched (does it ever need to check more than the first one)?

Side note, using "_" for $0 (the process name) in `bash -c` isn't very helpful.  I use "bash" for $0 because it is accurate, even if the rest of the arguments don't make much sense.
```suggestion
  for map in "$tDir"/stats/thresh_zstat*
  do
    mapCount=$(fslval "$map" dim4 | tr -d ' ')
    if ((mapCount > 1))
    then
      echo "warning: At least one component returned two z-score maps (alpha = 0.05 and 0.01)! Using first map only." 1>&2
      break
    fi
  done
```

Reply:
You can stop early, but you can't just check the first file. For test volume, the offending component is number 46 of 72.